### ISSUE TITLE
Fewer intersection matches in 4.1

### ISSUE THREAD

**Issue by mraross** [bcgov/ols-geocoder#192]
Investigate cause of reduction in intersection matches between 4.0 and 4.1 as reported by Jonathan Long at AG analyzing ICBC traffic ticket locations (see below). This could be related to issue #181

**Comment by mraross** [comment]
-------- Original message --------
From: "Longe, Jonathan AG:EX" <Jonathan.Longe@gov.bc.ca>
Date: 2020-12-23 11:46 AM (GMT-08:00)
To: "Kelsey, Brian CITZ:EX" <Brian.Kelsey@gov.bc.ca>, "Newall, Colin G CITZ:EX" <Colin.Newall@gov.bc.ca>, "Mounsey, Marli PSSG:EX" <Marli.Mounsey@gov.bc.ca>, "Wright, Eric A PSSG:EX" <Eric.Wright@gov.bc.ca>, "Schmill, Ulises PSSG:EX" <Ulises.Schmill@gov.bc.ca>, "Castel, Lex PSSG:EX" <Lex.Castel@gov.bc.ca>
Subject: Re: Geocoder 4.1 - Invitation to Test

Hey Brian and Colin,

Wow, good work DataBC!  Your new Geocoder is even better than than the current version.  Below are some stats that I generated by geocoding 2,680 addresses in both the new and current geocoder.

![image](https://user-images.githubusercontent.com/11318574/103163733-d2baa780-47b6-11eb-9ca0-4a588d8cb85c.png)

You can see that substantially more addresses have a precision of block or better using the new version.

I did find a couple of problem addresses that you may want to look into:

GREAT NORTHERN WAY AND KEITH DR, VANCOUVER, BC
DUNDAS AND WALL, VANCOUVER, BC
IRON WORKERS MEMORIAL BRIDGE, VANCOUVER, BC

Thanks again for the good work.  Please keep me posted on the roll-out.


Jonathan


From: Kelsey, Brian CITZ:EX
Sent: Wednesday, December 16, 2020 3:53:45 PM
Subject: Geocoder 4.1 - Invitation to Test
 
Invitation to Test Geocoder 4.1

DataBC has migrated version 4.1.0 of the BC Address Geocoder to the TEST environment. This migration includes improvements to address match accuracy. Your existing code should not be impacted by any of the changes.

Here is the What’s New page for the DataBC Home page. For more details, see the Detailed What's New page on GitHub. Here is a detailed list of Geocoder match accuracy improvements.

Please proceed with testing applications and scripts using the links below and let us know by January 7th if you encounter any issues. If everything goes as planned, we will release Geocoder 4.1 to production on January 26, 2021.

Online Geocoder

Test interactively using Location Services in Action (test version)
Send requests directly to the BC Address Online Geocoder (4.1) API in Test. Here is a sample request.
Batch Geocoder

Submit a job interactively using https://test.apps.gov.bc.ca/pub/cpf/secure/ws/apps/geocoder/#multiple
Submit a job using your batch list submitter script. In your script, to connect to the TEST environment, change the batch geocoder URL so that it contains ‘test.apps’ instead of ‘apps’ . Also, please let us know before you submit your first job so we can confirm that your CPF account has access.
Detailed Batch Geocoder UAT instructions can be found here:

https://github.com/bcgov/ols-geocoder/blob/gh-pages/batch-geocoder-uat.md

 

If you experience any issues please share with the Location Services team at: IDDDBCLS@Victoria1.gov.bc.ca

**Comment by mraross** [comment]
If you leave off the street types as in:

     main and first, cascade bc

you get the correct match:

     1st Ave and Main St, Cascade, BC 

with low confidence (e.g. a score of 88). There are two streetType.missing faults of 6 points each; if we reduce the penalty to 5 points, this match will get a score of 90.

**Comment by mraross** [comment]
Dundas and Wall, Vancouver, BC is the intersection of four streets:

     Dundas St and Powell St and Semlin Dr and Wall St, Vancouver, BC

so the match score of Dundas and Wall is quite low (72). Even providing all four street names without street types as follows:

     dundas and wall and powell and semlin , VANCOUVER, BC

returns a match score of 76. There are four streetType.missing penalties that, if reduced to five points would only improve the score to 80.

Would creating intersection aliases of pairs of streets for all intersections  involving more than two street names? So, the intersection with four names above would get six aliases as follows:

     Dundas St and Powell St , Vancouver, BC
     Dundas St and Semlin Dr, Vancouver, BC
     Dundas St and Wall St, Vancouver, BC
     Powell St and Semlin Dr, Vancouver, BC
     Powell St and Wall St, Vancouver, BC
     Semlin Dr and Wall St, Vancouver, BC

Street names that are generic need not be included in the aliases. For example, the following intersection:

     Douglas St and Gorge Rd E and Hillside Ave and turning lane, Victoria, BC

doesn't need aliases involving "turning lane". Just the following:

     Douglas St and Gorge Rd E and Hillside Ave, Victoria, BC
     Douglas St and Gorge Rd E, Victoria, BC
     Douglas St and Hillside Ave, Victoria, BC
     Gorge Rd E and Hillside Ave, Victoria, BC

The first alias may not even be required either.

[Here](https://geocodertst.api.gov.bc.ca/addresses.html?locationDescriptor=parcelPointmaxResults=10&addressString=dundas%20and%20wall,%20VANCOUVER,%20BC) is the request

**Comment by mraross** [comment]
ITN should be updated to change the following primary road segment names:

     Second Narrows Bridge, Vancouver, BC
     Second Narrows Bridge, District of North Vancouver, BC

to the following:

     Iron Workers Memorial Bridge, Vancouver, BC
     Iron Workers Memorial Bridge, District of North Vancouver, BC

Current primary names can be kept as aliases.

**Comment by gleeming** [comment]
For the intersection cases, we haven't done any testing or tweaking of those with the new garbage handling code.  It may be that the logic simply needs a slight modification to not-so-quickly toss out the word "and" before trying an intersection precision match.  If this doesn't produce the best results then an alias-based solution could be investigated.

With dundas and wall and powell and semlin , VANCOUVER, BC the v4.1 geocoder returns the same result as prod.  Perhaps a useful enhancement could be to have a new INTERSECTION_STREET_TYPE.missing fault with much lower penalty (e.g. 2 pts) instead of trying to tinker with a long-standing penalty for street matches.  This also makes more sense in terms of how a person might speak or enter a query: 1207 Douglas St (civic with type)  vs  Douglas and View (intersection with no types)

For the bridge case in your last comment, the official name is three words -- Ironworkers Memorial Bridge. This is already the primary name in ITN.

**Comment by mraross** [comment]
Great Northern Way doesn't intersect Keith Dr; it turns into E 6th Ave one block before it intersects Keith Dr.

![image](https://user-images.githubusercontent.com/11318574/103471989-7cacad80-4d3c-11eb-8a07-8017eeb4dedf.png)

The following intersection is what the user intended and returns a score of 100:

     E 6th Ave and Keith Dr, Vancouver, BC

**Comment by mraross** [comment]
@gleeming Great that Ironworkers Memorial Bridge is in there. All we need to add is the street name alias:

     Iron Workers Memorial Bridge

 to support how some of our clients actually enter that name.

**Comment by mraross** [comment]
@gleeming For how garbage removal affects intersection matching, see #181.

**Comment by mraross** [comment]
@gleeming Funny enough, I was thinking about IntersectionStreetType too.

**Comment by cmhodgson** [comment]
The STREET_NAME.notMatched penalty applied when some street names are not included or extra non-matching names are included in an intersection name, is also used for other cases (eg. if a site name match is found but the street name doesn't match). There should probably be a different penalty (maybe even 2 different, one for missing names and one for extras) for this case and it might have a reduced score.

Perhaps, INTERSECTION_STREET_NAME.notMatched, INTERSECTION_STREET_NAME.missing, and we could also change to INTERSECTION_STREET_TYPE.missing as well, but there are all the other street name penalties that would then be inconsistent unless we also changed them (for intersection matches only), eg. 
STREET_QUALIFIER.missing/notMatched/spelledWrong
STREET_DIRECTION.missing/notMatched/notMatchedInHighway/notPrefix/notSuffix/spelledWrong
STREET_NAME.spelledWrong/isAlias

**Comment by mraross** [comment]
@cmhodgson When you create intersection names, can you leave out generic street names like turning lane and yield lane. No one would ever use a generic name when referring to an intersection. So Douglas and Hillside becomes:

     Douglas St and Gorge Rd E and Hillside Ave

The exception would be If an intersection had only two street names and one of them was generic.

**Comment by cmhodgson** [comment]
I have removed the generic names as proposed above. I was concerned about intersections with more than two total street names where at least two of them are generic street names - I wanted to make sure that the decision of which generic name(s) to throw out and which to keep was consistent and deterministic. 
Eg. The intersection currently named "Patricia Bay Hwy and highway ramp and turning lane" could be reduced to either "Patricia Bay Hwy and highway ramp" OR "Patricia Bay Hwy and turning lane". To ensure this is consistent I chose to keep the alphabetically first generic names (if any generic names are to be kept at all).
